### PR TITLE
Clarify ``pure``.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Review by ekpyron — APPROVED** [review]
Well, it's an improvement...
I'm still wondering about ``msg.data``... for external calls, that's fine, but internal pure functions are really not pure anymore, if they read ``msg.data``...

In any case, we should then disallow ``codecopy`` and check if there's anything else that still violates this.

**Comment by chriseth** [comment]
Agreed! Can you create an issue and schedule for 0.9.x?

**Comment by ekpyron** [comment]
> Agreed! Can you create an issue and schedule for 0.9.x?

Do we have any way to actually schedule anything for 0.9.x? Maybe we should, but I wouldn't see any. Created https://github.com/ethereum/solidity/issues/12260 in any case.

File: docs/contracts/functions.rst [hunk 1]
```diff
@@ -222,6 +222,9 @@ Pure Functions
 --------------
 
 Functions can be declared ``pure`` in which case they promise not to read from or modify the state.
+In particular, it should be possible to evaluate a ``pure`` function at compile-time given
+only its inputs and ``msg.data``, but without any knowledge of the current blockchain state.
+This means that reading from ``immutable`` variables can be a non-pure operation.
```
Comment (axic) [inline docs/contracts/functions.rst:227]: Hmm, should we disallow `immutable` in `pure`? :)

Comment (axic) [reply docs/contracts/functions.rst:227]: Ok, it seems my review was stuck from a few hours ago ^^

But I see that now @ekpyron opened an issue for this.

Comment (ekpyron) [reply docs/contracts/functions.rst:227]: immutables are already view, I think that's what prompted the change here

Comment (axic) [reply docs/contracts/functions.rst:227]: Well than the text here is confusing as it suggests they are still allowed in pure.

Comment (ekpyron) [reply docs/contracts/functions.rst:227]: It does say it's a "non-pure operation", not sure how that's unclear :-)!

Comment (axic) [reply docs/contracts/functions.rst:227]: It says "**can be** a non-pure". So in some cases it "can be pure"?

Comment (ekpyron) [reply docs/contracts/functions.rst:227]: You mean it should have said ``is`` not ``can be`` :-)? Maybe, yeah...

Comment (ekpyron) [reply docs/contracts/functions.rst:227]: Actually, whenever reading an immutable **can be*** pure, it's actually a constant... except that we don't properly evaluate those :-D.

Comment (chriseth) [reply docs/contracts/functions.rst:227]: We do: https://github.com/ethereum/solidity/pull/10240

**Pull request merged**